### PR TITLE
Add checks for unknown interface calls/declarations in require blocks

### PR DESCRIPTION
**Pull request by cgzones**
* Add check for unknown interface call 
* Add check for declaration in require block not declared in own module

### CODE REVIEW

**Comment by dburgener** [comment]
My initial concern about W-010 is that currently if someone defines their own m4 macros and calls them, we support that just fine (they're just interface calls we don't know about).  I'm hesitant to warn on this scenario, because we would lose the ability to seamless support m4 macros.  Thoughts?

W-011 I like, but I think it should probably be a style check.  Seems analogous to S-002.  I can see the potential failure case (which S-002 has too), but the common case is that the declaration is to something that is to something in base.  I could be wrong, but it feels more like a style thing to me.

I haven't had a chance to read either of these in depth yet unfortunately.  I may have more comments later.

**Comment by cgzones** [comment]
**W-010**
We can allow specifying macros (full names or postfixes) in the configuration file
`assume_macros = { macro_one macro_two }`
`assume_macro_postfixes = { _macro _call }`

Or we only trigger if the call-name has a prefix, which matches a module name;
so we warn on `init_this-does-not-exist` but not on `can_dosomething` (assuiming there is no module named `can`).

**W-011**
I have no strong opinion whether this should be a warning or style check.
But I think there is one key difference to S-002:
If one uses a type from module bar as filecontext in module foo (foo.fc) and then tries to enable module foo without bar being enabled, libselinux will complain
```
libsepol.context_from_record: type bar_t is not defined
libsepol.context_from_record: could not create context structure
libsepol.context_from_string: could not create context structure
libsepol.sepol_context_to_sid: could not convert system_u:object_r:bar_t:s0 to sid
invalid context system_u:object_r:bar_t:s0
```

In contrast interface calls to non-base modules are generally enclosed in optional blocks.
So if one uses an interface of the non-base module foo, which requires a type from module bar, and module foo is enabled but bar not, then one might think the rules of this interface are active.
But they are not because the type from bar is not visible and type resolution failures aren't decisive in optional blocks.


p.s.: maybe the word *suffix* is more accurate than *postfix*?

**Comment by dburgener** [comment]
On W-010, I prefer the version where the prefix matches a module name.  I would accept a PR for that version of the check.

That's a strong argument for W-011 being a warning.  I agree.

I have no strong opinion on suffix vs postfix.  Whichever you prefer.

File: src/if_checks.c [hunk 1]
```diff
@@ -325,3 +325,62 @@ struct check_result *check_name_required_but_not_used_in_if(const struct
...
+		return NULL;
+	}
+
+	// TODO: handle templated declarations
+	if (name[0] == '$') {
+		return NULL;
+	}
+
+	// only check declarations in require blocks
+	for (const struct policy_node *cur = node; cur; cur = cur->parent) {
```
Comment (dburgener) [inline src/if_checks.c:352]: Is the interface/template bit essential?  Why not just use is_in_require() from tree.h?

**Comment by dburgener** [comment]
I'm also wavering on what I originally said about W-010 now that I've read the actual code for it.

Two issues with this approach are:
- If I have a typo in the prefix it won't get caught (eg domian_type())
- If someone defines m4 macros with a prefix matching an existing module, they have no way to determine that.

Maybe the other option you suggested would have been the better approach after all.  Thoughts?

**Comment by cgzones** [comment]
I liked the first version (scan all interface-call names) more.
With regard to refpolicy we would be fine with the exceptions `can_exec` and `*_pattern`.
For more broad compliance we might also ignore `can_*` and `gen_*`, or are you aware of policies using other named macros ?
I took a look at Androids sepolicy, they aren't using `interface`s at all but bare `define` macros.
As you mentioned you'd like to parse that in the future, we probably should then start to parse these files (e.g. https://android.googlesource.com/platform/system/sepolicy/+/refs/heads/master/public/te_macros) and then the macro names would be known to SELint.

**Comment by dburgener** [comment]
I'm aware of a non-public policy that has some macros with the same prefix as a module name, but situations like that are probably extremely rare.

I'm fine with merging this as it stands now and continuing to refine how we handle macros down the road.  People can always disable checks that aren't helpful in their situation.

If you want to take a look at my other comments in the changes, I'm happy to merge W-010 with the overall design now.  I think you're probably right that we'll want to parse macro files eventually, so we can revisit this check once we do that.

File: src/te_checks.c [hunk 2]
```diff
@@ -491,6 +491,68 @@ struct check_result *check_module_file_name_mismatch(const struct check_data
...
+		if (0 == strcmp(if_name, exception[i])) {
+			return NULL;
+		}
+	}
+
+	// ignore calls which does not start with a module name: they are probably macros
+	for (const char *prefix = strchr(if_name, '_'); prefix; prefix = strchr(prefix + 1, '_')) {
+		char *search_mod = strndup(if_name, (size_t)(prefix - if_name));
+
+		static const char *const abbreviations[][2] = {
```
[the remark this answers is not in the document]
Comment (cgzones) [reply src/te_checks.c]: moved into util.h

File: src/te_checks.c [hunk 3]
```diff
@@ -491,6 +491,68 @@ struct check_result *check_module_file_name_mismatch(const struct check_data
...
+		};
+		for (size_t i = 0; i < (sizeof abbreviations / sizeof *abbreviations); ++i) {
+			if (0 == strcmp(search_mod, abbreviations[i][0]) &&
+			    look_up_in_mods_map(abbreviations[i][1])) {
+				free(search_mod);
+				goto found_prefix;
+			}
+		}
+
+		if (look_up_in_mods_map(search_mod)) {
```
[the remark this answers is not in the document]
Comment (cgzones) [reply src/te_checks.c]: reordered

File: src/if_checks.c (hunk 1 above)
Comment (cgzones) [reply src/if_checks.c]: switched to `is_in_require()`

**Pull request merged**